### PR TITLE
wal: synchronize background operations

### PR DESCRIPTION
**Pull request by fabxc**
This adds an actor channel and thereby serializes all background
operations through the run() method.
Fixes an existing race.

@beorn7 @grobie

### CODE REVIEW

**Comment by beorn7** [comment]
I'm currently not in a state where I could do a qualified review, but I'll happily try out this patch on our test servers.

**Comment by beorn7** [comment]
I have built a Prom2 binary with this change (see commit 0de912895ff9973f032a20f6900ec7bcc9a6a906 in the prometheus/prometheus repo). At first, everything looked well, but after a day or two, two of our three servers went into some kind of stall. HTTP requests are still responded to, but there seem to be no ingestion or compaction taking place, almost no CPU usage. Nothing suspicious in the logs.

I haven't done any detailed investigation, and I don't know if this behavior has been introduced by this change, or we are just running into it now by bad luck. 

I have attached the goroutine dump of the two servers that appear to have stalled.


[prom2-collapsed-0de912895ff9973f032a20f6900ec7bcc9a6a906-stalled.20171029.txt](https://github.com/prometheus/tsdb/files/1425570/prom2-collapsed-0de912895ff9973f032a20f6900ec7bcc9a6a906-stalled.20171029.txt)

[prom2-collapsed-0de912895ff9973f032a20f6900ec7bcc9a6a906-stalled-2.20171029.txt](https://github.com/prometheus/tsdb/files/1425571/prom2-collapsed-0de912895ff9973f032a20f6900ec7bcc9a6a906-stalled-2.20171029.txt)

**Comment by fabxc** [comment]
Thanks. Yea, same deadlock in either coming from this PR. Basically we attempt to send on `actorc` while holding the WALs mutex. The processing loop started calling `Sync` though which waits for the mutex to be released. So `Sync` doesn't complete, and sending on `actorc` keeps blocking which means the mutex doesn't get released.

**Comment by fabxc** [comment]
Wrapped the channel sends in a goroutine again. This es essentially the same working behavior as before. But we still use the actor channel to serialize the operations in general.

**Comment by beorn7** [comment]
Running now on our test servers.
BTW: The test that's failing in Travis is passing for me locally.

**Comment by fabxc** [comment]
Yes, it's a flaky test (issue #163).

**Comment by beorn7** [comment]
FYI: This has now been running without any problem for the last 2d. Also, no races detected in the one server that ran with race detection enabled.

File: wal.go [hunk 1]
```diff
@@ -190,6 +190,7 @@ type SegmentWAL struct {
 
 	stopc   chan struct{}
 	donec   chan struct{}
+	actorc  chan func() error // sequantilized background operations
```
Comment (juliusv) [inline wal.go]: sequantilized -> sequentialized

**Comment by juliusv** [comment]
Looks good to me, but tests are still failing. Clicked on Retry in Travis just in case it's a flake...

**Pull request merged**